### DISCUSSION TITLE
Having problems with setup for an example from the docs

### DISCUSSION THREAD

**Discussion by shakeshuck** [gomlx/gomlx#464] [Q&A] [answered]
I'm having issues with dependencies when trying to cobble together examples from the docs on the site. I'm attempting to read in a dataset and use it to train an NN.
I'm a newbie to this, so please don't take anything for granted, I may have done something stupid!

In Visual Studio Code, I have the following imports:

	"github.com/gomlx/compute" // Includes default backends.
	"github.com/gomlx/gomlx/core/tensors"
	"github.com/gomlx/gomlx/ml/dataset"
	"github.com/gomlx/gomlx/ml/layers/activation"
	"github.com/gomlx/gomlx/ml/model"
	"github.com/gomlx/gomlx/ml/train"
	"github.com/gomlx/gomlx/ml/train/loss"
	"github.com/gomlx/gomlx/ml/train/optimizer"

but in this part of the example:

	modelFn := func(scope *model.Scope, spec any, inputs []*Node) []*Node {

		x := inputs[0] // shape: [batch_size, 2]

		h := denseLayer(scope.In("layer1"), x, 64)

		h = activation.Relu(h)

		h = denseLayer(scope.In("layer2"), h, 64)

		h = activation.Relu(h)

		h = denseLayer(scope.In("layer3"), h, 64)

		h = activation.Relu(h)

		// Output RGB values mapped to [0.0, 1.0] using Sigmoid

		y := Sigmoid(denseLayer(scope.In("output"), h, 3))

		return []*Node{y}

	}

the compiler is failing to find *Node or denseLayer definitions.

Am I missing a library, or is something else amiss?

Thanks.

**Comment by janpfeifer** [comment, marked as answer]
hi @shakeshuck ,

No worries. So the examples include the following import:

```go
import (
  ...
  . "github.com/gomlx/gomlx/core/graph"
  ...
)
```

This is a "dot import" (imports the symbols directly to the context of the file you are editing. This is generally not recommended in Go, but to write GoMLX computation graph functions, like the `modelFn`, we find it handy, so we use it. It will define the types `Graph`, `Node` and various operations like `Sigmoid`, `Add`, `Mul`, `Exp` that operate on graph `*Node` values.

The `denseLayer` I'm not sure where it came from (it's internal to the package you are editing). Notice that on the [_"First Model"_ ](https://gomlx.github.io/docs/first-model/) tutorial uses the `layers.Dense` library, which implements a NN Dense layer (typically a combination of a linear multiplication, a bias and then an activation).

I hope this helps. If not, I suggest following the [_"First Model"_  tutorial](https://gomlx.github.io/docs/first-model/), it should get you working, and then just change as it goes. The AI has been proficient in GoMLX for me as well, it can help -- but it can't replace your learning 😃

**Reply by shakeshuck** [reply]
Thanks Jan.

I found the original, it is in the "Get Started" section of your docs.
What I had missed, in my haste, was that the denseLayer function was written higher up the page.

I do have one more issue, though; the example contains:

```
	fmt.Println("Starting training loop...")

	_, err = loop.RunSteps(ds, *trainSteps)

	if err != nil {

		log.Fatalf("training failed: %v", err)

	}

```

...but there is no reference to trainSteps anywhere on the page that I can see?

**Reply by janpfeifer** [reply]
Here `trainSteps` refer to the number of training steps. And in some programs it is defined as a flag:

```go
var trainSteps = lag.Int("train_steps", 1000, "Number of steps to use during training.")
```

But you can use something else, however you like to pass the parameters. Most  of the demos take the `trainSteps` as a hyperparameter:

```go
trainSteps := model.GetParamOr(scope, "train_steps", 0)
```

(In this example `trainSteps`  is of type `int`, so there is no need to use `*` in front of it)

**Reply by shakeshuck** [reply]
Thanks again Jan.

I have a running example now.
